src/Lean/Parser/Term.lean

```diff
@@ -45,6 +45,7 @@ def seq1 :=
 end Tactic
 
 def darrow : Parser := " => "
+def semicolonOrLinebreak := ";" <|> checkLinebreakBefore >> pushNone
```

Comment:
Caveat: the current syntax pattern semantics make this case a bit peculiar: `` `(let $x:ident := $e; $f) `` will match regardless of whether `;` or a linebreak was used as tokens are ignored, but the same is not true when using a linebreak *in the pattern* as the `pushNone` will be checked structurally. We don't usually use linebreaks in syntax patterns at all though.

Reply:
> We don't usually use linebreaks in syntax patterns at all though.

https://github.com/leanprover-community/mathport/blob/37b4fd1c21239c9bda220c19208f88ff12f13d01/Mathport/Syntax/Transform/Tactic.lean#L94-L107 :smile: 

This is quite the gotcha (because it is also asymmetric).  Could we insert an empty atom instead of the null node?

If I understand you correctly, `sepByIndent` is not affected, right?  The pattern
```
{
$a
$b
}
```
will match all of:
`{a,b,}`, `{a,b}`,
```
{a
 b}
```
and
```
{a,
 b
}
```

Reply:
> This is quite the gotcha (because it is also asymmetric). Could we insert an empty atom instead of the null node?

I'm not sure I want to know what invariants an empty atom would break. Perhaps it's time to introduce a `matchIgnore` syntax kind or something we can put around the `<|>`.

> If I understand you correctly, sepByIndent is not affected, right?

It is affected in the same way; from the point of view of syntax `match`, there is no difference. In fact, ignoring the trailing separator presents an additional challenge. When matching a `nullNodeKind` pattern, syntax `match` currently eagerly checks that the number of elements in the pattern and in the discriminant is the same. We would have to change this such that a suffix of `matchIgnore` elements in the pattern may be missing from the discriminant :dizzy_face: .

Reply:
Ah I see, the `sepBy` matching issues are already present.  I mixed it up with the `$[$a],*` patterns, which ignore the separators.

> We would have to change this such that a suffix of `matchIgnore` elements in the pattern may be missing from the discriminant

We could also always generate a trailing `matchIgnore` node (which is either empty or contains the separator).

Reply:
@gebner Just to be sure, the mathport code already did not accept `;` before this PR, correct? If so, I will merge the PR for now before it bitrots and introduce `sepByIndent` later.